akka-stream/src/main/scala/akka/stream/impl/Throttle.scala

```diff
@@ -42,21 +41,27 @@ private[stream] class Throttle[T](cost: Int,
         else completeStage()
 
       override def onPush(): Unit = {
-        val timeElapsed = now() - previousTime
-        val currentTokens = Math.min(timeElapsed * speed + lastTokens, scaledMaximumBurst)
         val elem = grab(in)
         val elementCost = scale(costCalculation(elem))
-        if (currentTokens < elementCost)
-          mode match {
-            case Shaping ⇒
-              currentElement = Some(elem)
-              scheduleOnce(timerName, ((elementCost - currentTokens) / speed).millis)
-            case Enforcing ⇒ failStage(new RateExceededException("Maximum throttle throughput exceeded"))
-          }
-        else {
-          lastTokens = currentTokens - elementCost
-          previousTime = now()
+
+        if (lastTokens >= elementCost) {
+          lastTokens -= elementCost
           push(out, elem)
+        } else {
+          val currentTime = now()
+          val currentTokens = Math.min((currentTime - previousTime) * speed + lastTokens, scaledMaximumBurst)
+          if (currentTokens < elementCost)
+            mode match {
+              case Shaping ⇒
+                currentElement = Some(elem)
+                scheduleOnce(timerName, ((elementCost - currentTokens) / speed).nanos)
```

Comment:
since we didn’t update any values, this will not solve the problem for elements that cost more than the maximumBurst allows—these will be deferred indefinitely